src/Native/Runtime/arm/AllocFast.S

```diff
@@ -2,115 +2,470 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#include <unixasmmacros.inc>
-
 .syntax unified
 .thumb
 
-#ifdef _DEBUG
-.rodata
-AssertMsg:                    .asciz "__FILE__:%s: %s is not implemented\n"
-FileName:                     .asciz "AllocFast.S"
-RhpNewFastName:               .asciz "RhpNewFast"
-RhpNewFinalizableName:        .asciz "RhpNewFinalizable"
-RhpNewObjectName:             .asciz "RhpNewObject"
-RhpNewArrayName:              .asciz "RhpNewArray"
-RhpNewArrayRareName:          .asciz "RhpNewArrayRare"
-RhpNewFastAlign8Name:         .asciz "RhpNewFastAlign8"
-RhpNewFinalizableAlign8Name:  .asciz "RhpNewFinalizableAlign8"
-RhpNewFastMisalignName:       .asciz "RhpNewFastMisalign"
-RhpNewArrayAlign8Name:        .asciz "RhpNewArrayAlign8"
-
-.text
-.macro GEN_ASSERT_FUNC func
-        GEN_ASSERT AssertMsg, FileName, \func
-.endm
-#endif
-
-// TODO: Implement Arm support
+#include <unixasmmacros.inc>
+#include <AsmOffsets.inc>         // generated by the build from AsmOffsets.cpp
 
 // Allocate non-array, non-finalizable object. If the allocation doesn't fit into the current thread's
 // allocation context then automatically fallback to the slow allocation path.
 //  r0 == EEType
-//
-//    LEAF_ENTRY RhpNewFast, _TEXT
-//#ifdef _DEBUG
-//        GEN_ASSERT_FUNC RhpNewFastName
-//#endif
-//    LEAF_END RhpNewFast, _TEXT
-//
+LEAF_ENTRY RhpNewFast, _TEXT
+        PROLOG_PUSH "{r4,lr}"
+        mov	        r4, r0 // save EEType
+
+        // r0 = GetThread()
+        INLINE_GETTHREAD
+
+        mov         r1, r4 // restore EEType
+        // r1 contains EEType pointer
+        ldr         r2, [r1, #OFFSETOF__EEType__m_uBaseSize]
+
+        // r0: Thread pointer
+        // r1: EEType pointer
+        // r2: base size
+
+        ldr         r3, [r0, #OFFSETOF__Thread__m_alloc_context__alloc_ptr]
+        add         r2, r3
+        ldr         r4, [r0, #OFFSETOF__Thread__m_alloc_context__alloc_limit]
+        cmp         r2, r4
+        bhi         LOCAL_LABEL(RhpNewFast_RarePath)
+
+        // set the new alloc pointer
+        str         r2, [r0, #OFFSETOF__Thread__m_alloc_context__alloc_ptr]
+
+        mov         r0, r3
+
+        EPILOG_POP  "{r4,pc}"
+
+LOCAL_LABEL(RhpNewFast_RarePath):
+        EPILOG_POP  "{r4,lr}"
+        mov         r0, r1            // restore EEType
```

Comment:
It is more usual to do the tail call argument setup before epilog (ie before `EPILOG_POP`).